src/tbb/small_object_pool.cpp

```diff
@@ -40,6 +40,8 @@ void* __TBB_EXPORTED_FUNC allocate(d1::small_object_pool*& allocator, std::size_
     // TODO: optimize if the allocator contains a valid pool.
     auto tls = governor::get_thread_data();
     auto pool = tls->my_small_object_pool;
+    __TBB_ASSERT(allocator == nullptr || pool == allocator,
+                 "An attempt was made to allocate using another thread's small memory pool");
```

Comment:
I suggest moving this assert into `allocate_impl` function below, to put it just before line 68, where an assignment is made so that they become:
```cpp
__TBB_ASSERT(allocator == nullptr || allocator == this,
             "An attempt was made to allocate using another thread's small memory pool");
allocator = this;
```
This way the `allocate(small_object_pool, number_of_bytes, execution_data)` overload above is covered as well.

Reply:
IMHO, it would be better to place it in `allocate_impl` but before doing something, in the very beginning of the function